deadlines/src/test/java/com/palantir/deadlines/DeadlinesTest.java

```diff
@@ -0,0 +1,252 @@
+            long originalExternalValue = externalMeter.getCount();
+            long originalInternalValue = internalMeter.getCount();
+
+            DummyRequest outbound = new DummyRequest();
+            Deadlines.encodeToRequest(Duration.ofSeconds(10), outbound, new DummyRequestEncoder());
+
+            assertThat(externalMeter.getCount()).isGreaterThan(originalExternalValue);
+            assertThat(internalMeter.getCount()).isEqualTo(originalInternalValue);
+        }
+    }
+
+    @Test
+    public void test_encode_to_request_expiration_internal_deadline() {
+        try (CloseableTracer tracer = CloseableTracer.startSpan("test")) {
+            DummyRequest request = new DummyRequest();
+            Duration providedDeadline = Duration.ofMillis(100);
+            request.setHeader(
+                    DeadlinesHttpHeaders.EXPECT_WITHIN, Deadlines.durationToHeaderValue(providedDeadline.toNanos()));
+            Deadlines.parseFromRequest(Optional.of(Duration.ofMillis(1)), request, new DummyRequestDecoder());
+
+            // wait until we know the deadline has expired
+            Awaitility.pollInSameThread();
+            Awaitility.waitAtMost(Duration.ofSeconds(5)).untilAsserted(() -> {
+                Optional<Duration> remaining = Deadlines.getRemainingDeadline();
+                assertThat(remaining).hasValueSatisfying(d -> assertThat(d).isEqualTo(Duration.ZERO));
+            });
+
+            DeadlineMetrics metrics = DeadlineMetrics.of(SharedTaggedMetricRegistries.getSingleton());
+            Meter externalMeter = metrics.expired(Expired_Cause.EXTERNAL);
+            Meter internalMeter = metrics.expired(Expired_Cause.INTERNAL);
+            long originalExternalValue = externalMeter.getCount();
+            long originalInternalValue = internalMeter.getCount();
+
+            DummyRequest outbound = new DummyRequest();
+            Deadlines.encodeToRequest(Duration.ofSeconds(10), outbound, new DummyRequestEncoder());
+
+            assertThat(internalMeter.getCount()).isGreaterThan(originalInternalValue);
+            assertThat(externalMeter.getCount()).isEqualTo(originalExternalValue);
+        }
+    }
+
+    private static final class DummyRequest {
+        private final Multimap<String, String> headers = ArrayListMultimap.create();
+
+        void setHeader(String name, String value) {
+            headers.put(name, value);
+        }
+
+        Optional<String> getFirstHeader(String name) {
+            return headers.get(name).stream().findFirst();
+        }
+    }
+
+    private static final class DummyRequestEncoder implements RequestEncodingAdapter<DummyRequest> {
+        @Override
+        public void setHeader(DummyRequest dummyRequest, String headerName, String headerValue) {
+            dummyRequest.setHeader(headerName, headerValue);
+        }
+    }
+
+    private static final class DummyRequestDecoder implements RequestDecodingAdapter<DummyRequest> {
```

Comment:
might as well use enum singletons for these, no reason to make new instances.

I might consider deleting `DummyRequest` in favor of using `Multimap<String, String>` as the "request" type directly (or even Map<String, String> for simplicity)

Reply:
> might as well use enum singletons for these, no reason to make new instances.

done.

> I might consider deleting `DummyRequest` in favor of using `Multimap<String, String>` as the "request" type directly (or even Map<String, String> for simplicity)

I can do that, though I find `DummyRequest` to be a bit more readable personally. No strong preference though.

Reply:
ok fixed